libs/math/tests/unit/clustering/knn_test.cpp

```diff
@@ -18,41 +18,41 @@
 
 #include "math/clustering/knn.hpp"
 #include "math/distance/euclidean.hpp"
-#include "math/tensor.hpp"
+#include "test_types.hpp"
 
 #include "gtest/gtest.h"
 
-using namespace fetch::math;
-using namespace fetch::math::clustering;
+namespace fetch {
+namespace math {
+namespace test {
 
 template <typename T>
 class ClusteringTest : public ::testing::Test
 {
 };
 
-using MyTypes = ::testing::Types<fetch::math::Tensor<float>, fetch::math::Tensor<double>,
-                                 fetch::math::Tensor<fetch::fixed_point::FixedPoint<32, 32>>>;
-
-TYPED_TEST_CASE(ClusteringTest, MyTypes);
+TYPED_TEST_CASE(ClusteringTest, TensorFloatingTypes);
 
 TYPED_TEST(ClusteringTest, knn_euclidean_test)
 {
   using ArrayType = TypeParam;
+  using DataType  = typename TypeParam::Type;
   using SizeType  = typename TypeParam::SizeType;
 
   ArrayType A = ArrayType::FromString("1, 2, 3, 4; 2, 3, 4, 5; -1, -2, -3, -4; -2, -3, -4, -5");
   ArrayType v = ArrayType::FromString("3, 4, 5, 6");
 
-  auto output = KNN<ArrayType, fetch::math::distance::Euclidean>(A, v, 4);
+  auto output = clustering::KNN<ArrayType, fetch::math::distance::Euclidean>(A, v, 4);
 
   EXPECT_EQ(output.at(0).first, SizeType(1));
-  EXPECT_NEAR(double(output.at(0).second), double(2), 1e-4);
   EXPECT_EQ(output.at(1).first, SizeType(0));
-  EXPECT_NEAR(double(output.at(1).second), double(4), 1e-4);
   EXPECT_EQ(output.at(2).first, SizeType(2));
-  EXPECT_NEAR(double(output.at(2).second), double(14.6969384), 1e-4);
   EXPECT_EQ(output.at(3).first, SizeType(3));
-  EXPECT_NEAR(double(output.at(3).second), double(16.6132477), 1e-4);
+
+  EXPECT_TRUE(output.at(0).second - double(2) < function_tolerance<DataType>());
+  EXPECT_TRUE(output.at(1).second - double(4) < function_tolerance<DataType>());
+  EXPECT_TRUE(output.at(2).second - double(14.6973) < function_tolerance<DataType>());
+  EXPECT_TRUE(output.at(3).second - double(16.6132477) < function_tolerance<DataType>());
```

Comment:
Why has this changed from EXPECT_NEAR? The error messages from EXPECT_NEAR print both values and the tolerance which is helpful for debugging